modules/host/storageobligations.go

```diff
@@ -235,35 +364,383 @@ func (h *Host) addStorageObligation(so *storageObligation) error {
+	}
+	// Sanity check - all of the sector data should be modules.SectorSize
+	for _, data := range gainedSectorData {
+		if uint64(len(data)) != modules.SectorSize {
+			h.log.Critical("modifying a revision with garbase sector sizes", len(data))
+			return errInsaneStorageObligationRevision
+		}
+	}
+
+	// Note, for safe error handling, the operation order should be: add
+	// sectors, update database, remove sectors. If the adding or update fails,
+	// the added sectors should be removed and the storage obligation shoud be
+	// considered invalid. If the removing fails, this is okay, it's ignored
+	// and left to consistency checks and user actions to fix (will reduce host
+	// capacity, but will not inhibit the host's ability to submit storage
+	// proofs)
+	var i int
+	var err error
+	for i = range sectorsGained {
+		err = h.addSector(sectorsGained[i], so.expiration(), gainedSectorData[i])
+		if err != nil {
+			break
+		}
+	}
+	if err != nil {
+		// Because there was an error, all of the sectors that got added need
+		// to be revered.
+		for j := 0; j < i; j++ {
+			// Error is not checkeed because there's nothing useful that can be
+			// done about an error.
+			_ = h.removeSector(sectorsGained[j], so.expiration())
+		}
+		return err
+	}
+	// Update the database to contain the new storage obligation.
+	err = h.db.Update(func(tx *bolt.Tx) error {
+		soBytes, err := json.Marshal(*so)
 		if err != nil {
 			return err
 		}
+		return tx.Bucket(bucketStorageObligations).Put(soid[:], soBytes)
+	})
+	if err != nil {
+		// Because there was an error, all of the sectors that got added need
+		// to be revered.
+		for i := range sectorsGained {
+			// Error is not checkeed because there's nothing useful that can be
+			// done about an error.
+			_ = h.removeSector(sectorsGained[i], so.expiration())
+		}
+		return err
+	}
+	// Call removeSector for all of the sectors that have been removed.
+	for k := range sectorsRemoved {
+		// Error is not checkeed because there's nothing useful that can be
+		// done about an error. Failing to remove a sector is not a terrible
+		// place to be, especially if the host can run consistency checks.
+		_ = h.removeSector(sectorsRemoved[k], so.expiration())
+	}
+	return nil
+}
+
+// removeStorageObligation will remove a storage obligation from the host,
+// either due to failure or success.
+func (h *Host) removeStorageObligation(so *storageObligation, sos storageObligationStatus) error {
+	// Call removeSector for every sector in the storage obligation.
+	for _, root := range so.SectorRoots {
+		// Error is not checked, we want to call remove on every sector even if
+		// there are problems - disk health information will be updated.
+		_ = h.removeSector(root, so.expiration())
+	}
+
+	// Update the host revenue metrics based on the status of the obligation.
+	if sos == obligationConfused {
+		h.log.Critical("storage obligation confused!")
+	}
+	if sos == obligationRejected {
```

Comment:
`switch` is preferable because the values are exclusive (i.e. there's no possibility of fallthrough)